sui-execution/latest/sui-move-natives/src/eth_state_proof.rs

```diff
@@ -0,0 +1,265 @@
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, u64),
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, Vector).to_vec_u8()?,
+        pop_arg!(args, Vector).to_vec_u8()?,
+    );
+
+    let proof = bcs::from_bytes::<EIP1186ProofResponse>(proof.as_slice())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let message = String::from_utf8(message.clone())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let message_map_index = get_message_storage_slot(message.clone(), dwallet_id.clone(), map_slot)
+        .map_err(|_| PartialVMError::new(StatusCode::UNKNOWN_STATUS))?;
+
+    let contract_address = Address::from_slice(contract_address.as_slice());
+    let state_root = Bytes32::try_from(state_root.as_slice())
+        .map_err(|_| PartialVMError::new(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+
+    let account_path = keccak256(contract_address.as_bytes()).to_vec();
+    let account_encoded = encode_account(&proof);
+
+    // Verify the account proof against the state root.
+    let is_valid_account_proof = verify_proof(
+        &proof.clone().account_proof,
+        &state_root.as_slice().to_vec(),
+        &account_path,
+        &account_encoded,
+    );
+
+    if !is_valid_account_proof {
+        return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
+    }
+
+    // Get only the proof that matches the message_map_index.
+    let msg_storage_proof = proof
+        .storage_proof
+        .iter()
+        .find(|p| p.key == U256::from(message_map_index.as_bytes()))
+        .ok_or_else(|| PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR))?;
+    // Cast the storage key to a 32-byte array, and hash using keccak256 algorithm.
+    let mut msg_storage_proof_key_bytes = [0u8; 32];
+    msg_storage_proof
+        .key
+        .to_big_endian(&mut msg_storage_proof_key_bytes);
+
+    let storage_key_hash = keccak256(msg_storage_proof_key_bytes);
+    let storage_value = [TRUE_VALUE].to_vec();
+
+    let is_valid = verify_proof(
+        &msg_storage_proof.clone().proof,
+        &proof.storage_hash.as_bytes().to_vec(),
+        &storage_key_hash.to_vec(),
+        &storage_value,
+    );
+    Ok(NativeResult::ok(cost, smallvec![Value::bool(is_valid)]))
+}
+
+/***************************************************************************************************
+ * native fun verify_eth_state
+ * Implementation of the Move native function `eth_dwallet::verify_eth_state(proof: vector<vector<u8>>, proof_len: u64, root: vector<u8>, eth_smart_contract_address: vector<u8>, eth_smart_contract_slot: u64, message: vector<u8>): bool;`
+ * gas cost: verify_eth_state_cost_base   | base cost for function call and fixed operations.
+ **************************************************************************************************/
```

Comment:
```suggestion
/***************************************************************************************************
 * native fun verify_eth_state
 * Implementation of the Move native function
 * `eth_dwallet::verify_eth_state(proof: vector<vector<u8>>,
 * proof_len: u64,
 * root: vector<u8>,
 * eth_smart_contract_address: vector<u8>,
 * eth_smart_contract_slot: u64,
 * message: vector<u8>): bool;`
 * gas cost: verify_eth_state_cost_base   | base cost for function call and fixed operations.
 **************************************************************************************************/
```
Also, the params are wrong.